### ISSUE TITLE
Lag of the video stream from real time. Problem of handling errors when disconnecting the camera during processing.

### ISSUE THREAD

**Issue by lubagov** [vignetteapp/SeeShark#29]
Hello! I'm interested in this code and the asynchronous processing option.

SeeShark\SeeShark\Device\VideoDevice.cs:
```
        public DecodeStatus TryGetFrame(out Frame frame)
        {
            DecodeStatus status = decoder.TryDecodeNextFrame(out frame);

            // Big brain move to avoid overloading the CPU \o/
            // Decide whether we wait longer during a Thread.Sleep() when there are no frames available.
            // Waiting longer would mean a full frame interval (for example ~16ms when 60 fps), 1ms otherwise.
            // Always wait longer just after receiving a new frame.
            bool waitLonger = status == DecodeStatus.NewFrame;
            Thread.Sleep(waitLonger ? 1000 * decoder.Framerate.den / (decoder.Framerate.num + 5) : 1);

            return status;
        }
```
And so is this loop:
I append here OnEndOfStream event:
```
        public event EventHandler<DecodeStatus>? OnEndOfStream;
. . . 
        protected void DecodeLoop()
        {
            DecodeStatus status;
            while ((status = TryGetFrame(out var frame)) != DecodeStatus.EndOfStream)
            {
                OnFrame?.Invoke(this, new FrameEventArgs(frame, status));

                if (!IsPlaying)
                    break;
            }
            if (status == DecodeStatus.EndOfStream)
            {
                OnEndOfStream?.Invoke(this, status);
            }
        }
```

I slightly modified the code here, adding a handler in case of an error. 

The problem is that in case of an error, if the camera was disconnected from USB during processing, calling `ffmpeg.av_read_frame(FormatContext, Packet);` will return an `EIO` error with code -5:

SeeShark\SeeShark\Decode\VideoStreamDecoder.cs:
```
                error = ffmpeg.av_read_frame(FormatContext, Packet);

                if (error < 0)
                {
                    nextFrame = Frame;
                    GC.Collect();

                    // We only wait longer once to make sure we catch the frame on time.
                    return error == eagain
                        ? DecodeStatus.NoFrameAvailable
                        : DecodeStatus.EndOfStream;
                }
```
EIO is not the same as EAGAIN and will return DecodeStatus.EndOfStream accordingly. TryGetFrame will return EndOfStream along the chain, then DecodeLoop will exit the loop and **IsPlaying will remain True** (here we not set it to false on exit from loop), but we won't get any more frames. We also don't know about the error. those. **OnFrame?.Invoke will not be called**, with the status DecodeStatus.EndOfStream and there **is no way to determine that an error has occurred** and the camera is no longer working.

To solve this problem i append here `OnEndOfStream` and process it here. Other option still call OnFrame again with this status. But as it is now, we simply cannot find out that the camera has stopped, due to an unexpected error..

Problem two...
```
            bool waitLonger = status == DecodeStatus.NewFrame;
            Thread.Sleep(waitLonger ? 1000 * decoder.Framerate.den / (decoder.Framerate.num + 5) : 1);
```
This loop has a delay based on the FPS of the camera, which it declares in its configuration. Many cameras are able to adapt to light levels, I found this on both the ELP IMX 179 camera and the Genius camera. The fact is that, regardless of the FPS specified in the settings, they can raise it or lower it, compensating for too dark or overexposed frames.

I tried lighting the camera with a direct light with a lamp, and it raised the FPS to over 30 fps.
Over time, I got a video lag from reality, because each successfully received frame waited for the full time to the FPS indicated by the camera.

In case of low light, FPS drops to 14FPS or even lower. As a result, the loop executes several times with a delay of 1ms, which is also not good...

One more thing, if there is at least some processing in the `OnFrame` event, then this will lead to the same lag from reality. Because the event runs on the same thread as the loop, and processing time will increase the overall total delay between frames.

i append here next code, to look what happenes here:
```
        public DecodeStatus TryGetFrame(out Frame frame)
        {
            DecodeStatus status = decoder.TryDecodeNextFrame(out frame);
            bool waitLonger = status == DecodeStatus.NewFrame;
            int waitTime = waitLonger ? 1000 * decoder.Framerate.den / (decoder.Framerate.num + 5) : 1;
            Console.WriteLine($"wait time: {waitTime} {decoder.Framerate.den} {decoder.Framerate.num}");
            Thread.Sleep(waitTime);
            return status;
        }
```
the output, when camera increase FPS:
```
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
NewFrame
wait time: 33 333333 10000000
```
the output, when camera reduce FPS:
```
NewFrame
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 1 333333 10000000
NoFrameAvailable
wait time: 33 333333 10000000
NewFrame
```

Thank you!

**Comment by Speykious** [comment]
Thanks for your feedback! I indeed completely neglected to properly handle end of stream in the video stream decoder.

As for everything else, especially the decode loop, do you think you can do a PR about it? Seeing some actual implementation would really help us

**Comment by Speykious** [comment]
> This loop has a delay based on the FPS of the camera, which it declares in its configuration. Many cameras are able to adapt to light levels, I found this on both the ELP IMX 179 camera and the Genius camera. The fact is that, regardless of the FPS specified in the settings, they can raise it or lower it, compensating for too dark or overexposed frames.
> 
> I tried lighting the camera with a direct light with a lamp, and it raised the FPS to over 30 fps.
> Over time, I got a video lag from reality, because each successfully received frame waited for the full time to the FPS indicated by the camera.
> 
> In case of low light, FPS drops to 14FPS or even lower. As a result, the loop executes several times with a delay of 1ms, which is also not good...

Can reproduce with my OBS virtual camera by increasing the FPS after launching `SeeShark.Example.Stats`.
I'll have to see if I can make the FPS update in real time as well.

**Comment by Speykious** [comment]
> One more thing, if there is at least some processing in the `OnFrame` event, then this will lead to the same lag from reality. Because the event runs on the same thread as the loop, and processing time will increase the overall total delay between frames.

That is true. However, the solution would involve some copying of the received frame data to make it thread-safe, which could produce some performance drop - or at least consume more memory and CPU per frame. I'll have to investigate.

**Comment by lubagov** [comment]
> That is true. However, the solution would involve some copying of the received frame data to make it thread-safe, which could produce some performance drop - or at least consume more memory and CPU per frame. I'll have to investigate.



```
        private void OnFrameEventHandler(object sender, FrameEventArgs e)
        {
            var frame = e.Frame;

            if ((frame.Width == 0) || (frame.Height == 0) || (e.Status==DecodeStatus.NoFrameAvailable))
                return;

            converter ??= new FrameConverter(frame, PixelFormat.Rgb24);

            lock (syncPrimitive)
            {
                if (CurFrame != null)
                {
                    return;
                }

                Frame cFrame = converter.Convert(frame);
                CurFrame = new ImageFrame(ImageFormat.Srgb,
                    cFrame.Width, cFrame.Height, cFrame.WidthStep, cFrame.RawData);
                
                Monitor.Pulse(syncPrimitive);
            }
        }
```

I'm currently working with it like this. This is not efficient, but for a small resolution of 640x480 it will do. I don't need more now.

In my experience, the biggest waste of time is not in copying, but in memory allocation. It takes several times more time to allocate a buffer for a frame, if, say, this is not a compressed 8 megapixel video than to copy. The converter allocates memory every time, every time it is called, for every new frame inside, and I cannot reuse the buffer.

In my experience, it's more efficient to have a pool of buffers with memory allocated once. Remove one buffer from the pool, use it for video decoding, and give it to the user. The user can return it to the pool as needed (reducing the number of refs).
Without copying the frame, inside the OnFrameEventHandler (using only a reference), this doesn't seem to work either. after the frame is destroyed.
Maybe I'm wrong about something.

**Comment by sr229** [comment]
Going to delegate this on medium priority, as we do need to investigate this but not immediately.

**Comment by lubagov** [comment]
> Going to delegate this on medium priority, as we do need to investigate this but not immediately.

Since in this form, the library cannot be used in production. Because this inevitably leads to frame lag at nominal (due to the time spent copying the frame) or increased FPS.

I need to look for some quick solution, or go to other libraries. For example, work with the camera through OpenCV.
In my particular task, the stability of frames and the intervals between them is not important, then it is permissible to wait only if the frame is not decoded. For example 1/4 of FPS. And don't wait at all if the frame is decoded successfully to prevent lag. The question is what is there to decode synchronously in FFMPEG .... so that we know how many more frames it buffered.

**Comment by Speykious** [comment]
> The converter allocates memory every time, every time it is called, for every new frame inside, and I cannot reuse the buffer.

Actually no, the `FrameConverter` only allocates a frame buffer for the destination frame once in its constructor:
https://github.com/vignetteapp/SeeShark/blob/7768179eac3cb6095823610859b566e33c8589d2/SeeShark/FrameConverter.cs#L80-L94
```cs
var convertedFrameBufferSize = ffmpeg.av_image_get_buffer_size(dstPF, dstWidth, dstHeight, 1);
convertedFrameBufferPtr = Marshal.AllocHGlobal(convertedFrameBufferSize);

var dstData = new byte_ptrArray4();
var dstLinesize = new int_array4();
ffmpeg.av_image_fill_arrays(ref dstData, ref dstLinesize,
    (byte*)convertedFrameBufferPtr, dstPF, dstWidth, dstHeight, 1);

var dstFrame = ffmpeg.av_frame_alloc();
dstFrame->width = DstWidth;
dstFrame->height = DstHeight;
dstFrame->data.UpdateFrom(dstData);
dstFrame->linesize.UpdateFrom(dstLinesize);
dstFrame->format = (int)dstPF;
DstFrame = new Frame(dstFrame);
```

And all the `Convert` method does is to call `sws_scale` and return the already existing `DstFrame`, so it does not allocate more.
https://github.com/vignetteapp/SeeShark/blob/7768179eac3cb6095823610859b566e33c8589d2/SeeShark/FrameConverter.cs#L97-L106
```cs
public Frame Convert(Frame srcFrame)
{
    var srcAVFrame = srcFrame.AVFrame;
    var dstAVFrame = DstFrame.AVFrame;
    ffmpeg.sws_scale(convertContext,
        srcAVFrame->data, srcAVFrame->linesize, 0, srcAVFrame->height,
        dstAVFrame->data, dstAVFrame->linesize);

    return DstFrame;
}
```

This would need to be changed if we end up using a pool of buffers instead of just one.
The same is true for `VideoStreamDecoder`, it does not allocate any additional buffer every frame, which is why SeeShark cannot share `Frame`s between threads safely.

**Comment by Speykious** [comment]
> In my particular task, the stability of frames and the intervals between them is not important, then it is permissible to wait only if the frame is not decoded. For example 1/4 of FPS. And don't wait at all if the frame is decoded successfully to prevent lag.

Looking at the `TryGetFrame` method, I understand what you mean:
https://github.com/vignetteapp/SeeShark/blob/7768179eac3cb6095823610859b566e33c8589d2/SeeShark/Device/VideoDevice.cs#L72-L84

A better solution would be to wait *before* decoding the next frame. However, not waiting at all will consume 100% of a CPU thread, that I cannot allow.
You do bring another good solution, which is to wait for a certain fraction of the FPS instead. I'll try that.

Yet again though, if you think you can make a pull request for this enhancement, don't hesitate. :)

**Comment by lubagov** [comment]
>> cannot share `Frame`s between threads safely.

That's just the point... And I was able to pull the frame out of this thread, only after copying it, with the help of `Mediapipe.Net.Framework.Format.ImageFrame`.  Which allocated new memory, and created a new instance of mediapipe::ImageFrame internally. And then ImageFramePacket called Dispose on it, after processing... And it couldn't be reused. https://github.com/vignetteapp/MediaPipe.NET/blob/d562895694bb9d590b328682bf869c51eec24205/Mediapipe.Net/Framework/Packets/PacketFactory.cs#L84

`A better solution would be to wait before decoding the next frame. However, not waiting at all will consume 100% of a CPU thread, that I cannot allow.`

What's the difference when to wait ... Before or after ... The main thing is that if we wait when there are a lot of frames in the buffer ... then we will not catch up with them. And if we wait when there are frames in the buffer, and then further increase the waiting time by the processing or copying time, in the same stream... We will lag behind even more.

**Comment by lubagov** [comment]
Some like this:

```
diff --git a/SeeShark/Device/VideoDevice.cs b/SeeShark/Device/VideoDevice.cs
index 22e8e75..912cb92 100644
--- a/SeeShark/Device/VideoDevice.cs
+++ b/SeeShark/Device/VideoDevice.cs
@@ -17,6 +17,7 @@ namespace SeeShark.Device
         public bool IsPlaying { get; private set; }
 
         public event EventHandler<FrameEventArgs>? OnFrame;
+        public event EventHandler<DecodeStatus>? OnEndOfStream;
 
         public VideoDevice(VideoDeviceInfo info, DeviceInputFormat inputFormat, VideoInputOptions? options = null)
         {
@@ -34,6 +35,11 @@ namespace SeeShark.Device
                 if (!IsPlaying)
                     break;
             }
+            if (status == DecodeStatus.EndOfStream)
+            {
+                OnEndOfStream?.Invoke(this, status);
+            }
+            IsPlaying = false;
         }
 
         /// <summary>
@@ -77,9 +83,10 @@ namespace SeeShark.Device
             // Decide whether we wait longer during a Thread.Sleep() when there are no frames available.
             // Waiting longer would mean a full frame interval (for example ~16ms when 60 fps), 1ms otherwise.
             // Always wait longer just after receiving a new frame.
-            bool waitLonger = status == DecodeStatus.NewFrame;
-            Thread.Sleep(waitLonger ? 1000 * decoder.Framerate.den / (decoder.Framerate.num + 5) : 1);
-
+            if (status == DecodeStatus.NoFrameAvailable)
+            {
+                Thread.Sleep(1000 * decoder.Framerate.den / decoder.Framerate.num / 2);
+            }
             return status;
         }
 
```

those. instead of waiting when a frame is available, wait when it's not available, and only half the time. So that there is time to copy it when it is available.

**Comment by Speykious** [comment]
> What's the difference when to wait ... Before or after ... The main thing is that if we wait when there are a lot of frames in the buffer ... then we will not catch up with them.

FFmpeg, and thus by extension SeeShark, **does not buffer frames at all**. Only MediaPipe does.
You can test this with a loop where you call `camera.GetFrame()` and then wait for a little while with `Thread.Sleep(500)` on a camera with high FPS. You will see that it fetches the latest image of the camera. It never lags behind.

This is why it makes a difference. Fetching the image and then waiting means that we're having a slight delay from the camera to the code. Waiting and then fetching the image removes that artificial delay.

**Comment by lubagov** [comment]
It is hard to say. But also work with MediaPipe in a separate thread. When I shine a lamp on the camera, I get a delay. Gradually, it accumulates, makes up, let's say, a minute ... Then, obviously, everything will collapse after a while, when the memory runs out.

**Comment by Speykious** [comment]
Make a test with your camera using only SeeShark and not MediaPipe so that you can make sure that SeeShark is not at fault here. I'm certain that the FFmpeg C API does not buffer frames internally.

**Comment by lubagov** [comment]
Well... maybe it's a feature in Windows? But yes, I did tests without MediaPipe... And again, I don't call MediaPipe on the camera thread.

**Comment by adryzz** [comment]
might be... from my testing months ago on linux using a virtual camera at various fps (i tested 1-240fps) it didn't slow down at all.

can't confirm that right now as I'm not home this week.

**Comment by Speykious** [comment]
> Some like this:
> 
> ```
> diff --git a/SeeShark/Device/VideoDevice.cs b/SeeShark/Device/VideoDevice.cs
> index 22e8e75..912cb92 100644
> --- a/SeeShark/Device/VideoDevice.cs
> +++ b/SeeShark/Device/VideoDevice.cs
> @@ -17,6 +17,7 @@ namespace SeeShark.Device
>          public bool IsPlaying { get; private set; }
>  
>          public event EventHandler<FrameEventArgs>? OnFrame;
> +        public event EventHandler<DecodeStatus>? OnEndOfStream;
>  
>          public VideoDevice(VideoDeviceInfo info, DeviceInputFormat inputFormat, VideoInputOptions? options = null)
>          {
> @@ -34,6 +35,11 @@ namespace SeeShark.Device
>                  if (!IsPlaying)
>                      break;
>              }
> +            if (status == DecodeStatus.EndOfStream)
> +            {
> +                OnEndOfStream?.Invoke(this, status);
> +            }
> +            IsPlaying = false;
>          }
>  
>          /// <summary>
> @@ -77,9 +83,10 @@ namespace SeeShark.Device
>              // Decide whether we wait longer during a Thread.Sleep() when there are no frames available.
>              // Waiting longer would mean a full frame interval (for example ~16ms when 60 fps), 1ms otherwise.
>              // Always wait longer just after receiving a new frame.
> -            bool waitLonger = status == DecodeStatus.NewFrame;
> -            Thread.Sleep(waitLonger ? 1000 * decoder.Framerate.den / (decoder.Framerate.num + 5) : 1);
> -
> +            if (status == DecodeStatus.NoFrameAvailable)
> +            {
> +                Thread.Sleep(1000 * decoder.Framerate.den / decoder.Framerate.num / 2);
> +            }
>              return status;
>          }
>  
> ```
> 
> those. instead of waiting when a frame is available, wait when it's not available, and only half the time. So that there is time to copy it when it is available.

Are you talking about those changes? If that's the case, try my suggestion to wait before fetching the frame instead of after (with the current code for waiting instead of your modification for testing purposes). Normally, you shouldn't have a significant delay between the frame of the camera and what SeeShark receives.

**Comment by lubagov** [comment]
It seems to me that it does not matter at all when to wait, before receiving a frame or after ... When we wait 1/30 of a second and the delay is 60 seconds ...

**Comment by adryzz** [comment]
I have an idea... as i currently do not own a camera and am planning to get one, could you please point me to exactly which camera you have, so that i can get one and test this too? the ELP IMX 179, as i can see multiple versions of this one.

**Comment by lubagov** [comment]
Some like this. https://aliexpress.ru/item/1005004562138129.html?item_id=1005004562138129&sku_id=12000029625689254&spm=a2g2w.productlist.search_results.12.504b5644JrUaNQ
 It's hard to say, it was bought more than 5 years ago, for my previous job. I can't say that this is a good camera. Usually it runs 15FPS, although it declares 640x480 30 FPS. under normal room lighting. (However, in bright light in the Shopping Mall, such cameras increase FPS.). I sent a candle lamp to the camera to increase the FPS.

![cam_shot](https://user-images.githubusercontent.com/6134203/181013146-eac0008c-c620-421c-a548-6898ea53ab33.png)

But this problem seems to be easier to reproduce ... inside, camera.OnFrame, put, say, Thread.sleep (1000), then remove it, let's say 60 times. And then it will all display frames with a lag of 60 seconds.

**Comment by lubagov** [comment]
@Speykious I guess I figured out how to show it. Those changes that I made in the code are applied. So the process does not wait between frames.
Test code:
```
// Copyright (c) The Vignette Authors
// This file is part of SeeShark.
// SeeShark is licensed under the BSD 3-Clause License. See LICENSE for details.

using FFmpeg.AutoGen;
using SeeShark;
using SeeShark.Decode;
using SeeShark.Device;
using SeeShark.FFmpeg;


namespace Main
{
    class Program
    {

        int i = 0;
        long lastTime = DateTime.Now.Ticks;

        void onEndOfStream(object? sender, DecodeStatus e)
        {
            Console.WriteLine(e);
        }


        private void onFrameEventHandler(object? sender, FrameEventArgs e)
        {
            var frame = e.Frame;
            if ((frame.Width == 0) || (frame.Height == 0) || (e.Status == DecodeStatus.NoFrameAvailable))
                return;

            if ((i > 200) && (i < 300))
            {
                Thread.Sleep(200);
            }
            long nowTime = DateTime.Now.Ticks;
            double iv = (nowTime - lastTime) / (double)TimeSpan.TicksPerSecond;
            lastTime = nowTime;
            Console.WriteLine($"Frame number: {i}, Sec delay between frames:  {iv}");
            ++i;
        }

        public Program()
        {
            Console.WriteLine("Hello, World!");

            FFmpegManager.SetupFFmpeg(@"C:\ffmpeg\v5.0_x64\bin\", "/usr/lib");

            (int, int)? videoSize = (640, 480);
            Camera? camera = null;
            try
            {

                using (CameraManager manager = new CameraManager())
                {
                    camera = manager.GetDevice(0,

                        new VideoInputOptions
                        {
                            InputFormat = null,
                            Framerate = new AVRational
                            {
                                num = (int)30,
                                den = 1,
                            },
                            VideoSize = videoSize,
                        }
                    );
                    Console.WriteLine($"Using camera {camera.Info}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"cannot initialize camera {ex.Message}");
                return;
            }

            camera.OnFrame += onFrameEventHandler;
            camera.OnEndOfStream += onEndOfStream;

            camera.StartCapture();
            Thread.Sleep(1000000);
            camera.StopCapture();
            Console.WriteLine("Camera stopped!");
            camera?.Dispose();
            GC.Collect();
            GC.WaitForPendingFinalizers();
        }

        public static void Main(String[] args)
        {
            new Program();
        }
    }
}
```
Output:
[output.txt](https://github.com/vignetteapp/SeeShark/files/9192172/output.txt)

And llok on output after 300 frame, before we have delay 0.2 sec, next we at all have not delay:
```
Frame number: 294, Sec delay between frames:  0,2029678
Frame number: 295, Sec delay between frames:  0,2029045
Frame number: 296, Sec delay between frames:  0,2033898
Frame number: 297, Sec delay between frames:  0,2051678
Frame number: 298, Sec delay between frames:  0,2025567
Frame number: 299, Sec delay between frames:  0,2047621
Frame number: 300, Sec delay between frames:  0,007351
Frame number: 301, Sec delay between frames:  0,0048417
Frame number: 302, Sec delay between frames:  0,0061096
Frame number: 303, Sec delay between frames:  0,0088672
Frame number: 304, Sec delay between frames:  0,0043009
Frame number: 305, Sec delay between frames:  0,0030255
Frame number: 306, Sec delay between frames:  0,0025584
Frame number: 307, Sec delay between frames:  0,0026753
```
and after 575 frame we get normal FPS back:
```
Frame number: 573, Sec delay between frames:  0,0013264
Frame number: 574, Sec delay between frames:  0,0013013
Frame number: 575, Sec delay between frames:  0,0178284
Frame number: 576, Sec delay between frames:  0,0364166
Frame number: 577, Sec delay between frames:  0,0342118
Frame number: 578, Sec delay between frames:  0,0341714
```

**Comment by Speykious** [comment]
How was the camera FPS during this test?

**Comment by lubagov** [comment]
I illuminated the camera with a light bulb directed. So the time between frames was about 0.03 seconds. This can be seen up to frame 200. Then I slowed down the camera with a 0.2 second delay between frames, 99 times. those. I did 19.8 seconds delay. During this time, 276 frames have accumulated somewhere. And it threw them out very quickly, when there was no delay. In the interval from 300 to 576 frames.
(Although on the other hand, 570 frames should have accumulated in 19 seconds at 30FPS)

If you make a delay between frames, as it was without my changes, this delay does not allow you to quickly discard the accumulated frames ... And I observe the image lag.

**Comment by Speykious** [comment]
I tested your small modification (without the end of stream event) with my webcam on Linux, and actually observed some catching up as well, which was quite surprising.
![image](https://user-images.githubusercontent.com/13885008/181089517-66343f09-8fe0-4e5e-8fdd-d1179a48b79b.png)

Something interesting to note is that I applied your code to our `SeeShark.Example.Ascii` example, which revealed some amount of buffering with my webcam for a few seconds before eventually giving up and just send its last frames, or should I say the last frame of its limited buffer (all during frame numbers 200 to 300).
I don't know which part of this whole process is buffering, I just know that it is definitely not FFmpeg, and now it seems to depend on the camera or driver.
That's why I can't accurately reproduce your issue, because it might come from the way Windows handles it.

Regardless, thank you for your test, now I understand what's going on.
This is a very simple temporary fix while waiting for SeeShark to send frames on a different thread.
